### PR TITLE
Avoid possible divide-by-zero in Vectors layer thumbnail update

### PR DESCRIPTION
**Pull request by aganders3**
# Description

Trying to run tests locally I ran into nine tests in `layers/vectors/_tests/test_vectors.py` tests consistently failing:
```
FAILED napari/layers/vectors/_tests/test_vectors.py::test_no_args_vectors - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_no_data_vectors_with_ndim - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_empty_vectors - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_empty_vectors_with_property_choices - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_empty_layer_with_edge_colormap - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_empty_layer_with_edge_color_cycle - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_no_data_3D_vectors_with_ndim - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_empty_3D_vectors - RuntimeWarning: divide by zero encountered in divide
FAILED napari/layers/vectors/_tests/test_vectors.py::test_empty_data_from_tuple - RuntimeWarning: divide by zero encountered in divide
```

I will spare the repeated output because it was basically the same for each:
```
_________________________________________ test_no_args_vectors _________________________________________

    def test_no_args_vectors():
        """Test instantiating Vectors layer with no arguments"""
>       layer = Vectors()

napari/layers/vectors/_tests/test_vectors.py:40: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
napari/layers/vectors/vectors.py:248: in __init__
    self._update_dims()
napari/layers/base/base.py:704: in _update_dims
    self.refresh()  # This call is need for invalidate cache of extent in LayerList. If you remove it pleas ad another workaround.
napari/layers/base/base.py:1210: in refresh
    self._update_thumbnail()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Vectors layer 'Vectors' at 0x287fdd7c0>

    def _update_thumbnail(self):
        """Update thumbnail with current vectors and colors."""
        # calculate min vals for the vertices and pad with 0.5
        # the offset is needed to ensure that the top left corner of the
        # vectors corresponds to the top left corner of the thumbnail
        de = self._extent_data
        offset = (np.array([de[0, d] for d in self._dims_displayed]) + 0.5)[
            -2:
        ]
        # calculate range of values for the vertices and pad with 1
        # padding ensures the entire vector can be represented in the thumbnail
        # without getting clipped
        shape = np.ceil(
            [de[1, d] - de[0, d] + 1 for d in self._dims_displayed]
        ).astype(int)[-2:]
>       zoom_factor = np.divide(self._thumbnail_shape[:2], shape).min()
E       RuntimeWarning: divide by zero encountered in divide

napari/layers/vectors/vectors.py:682: RuntimeWarning
```

The root of the problem is the undefined behavior in casting a `NaN` to `int` because `self._extent_data` is filled with `NaN` when the data is empty.
```
# natively on my Apple Silicon Mac
>>> np.float32("nan").astype(int)
0

# in an (emulated) x86-64 Ubuntu container on my machine
>>> np.float32("nan").astype(int)
-9223372036854775808
```

It looks like this or something like it was fixed or anticipated for the Shapes layer, so I copied the implementation style from there. I checked some of the other layer types and they look okay as well, though I'm still getting familiar with them.

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
See https://github.com/numpy/numpy/issues/14412
I think this is actually set to emit a warning in a coming version of numpy, and potentially raise an error.

# How has this been tested?
Vector tests pass with my change on multiple platforms.

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- N/A I have made corresponding changes to the documentation
- N/A I have added tests that prove my fix is effective or that my feature works
- N/A If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

**Comment by psobolewskiPhD** [comment]
Thanks for addressing this!
Nice detective work as to the real cause and why it passes on CI but crops up on arm64 macs!
We'd wondered about it before: 
https://napari.zulipchat.com/#narrow/stream/212875-general/topic/failing.20test.20on.20latest.20main

Tested locally, all vector tests on my M1 pass and
`viewer.add_vectors()` no longer results in 3 warnings.
👍 

For anyone reviewing, the diff is hard to follow, no idea why it's generated like that. Easier to look here perhaps:
https://github.com/napari/napari/blob/dddbdcf39479073a693dc2ad1e989f371f62b6aa/napari/layers/vectors/vectors.py#L667-L674
Those are the added lines, with the two `colormapped` lines being moved from 699-700. Then the rest is moved into the `else:` as is, with no changes but the indentation and black formatting.
VS Code does a better job of presenting it:
image

**Comment by brisvag** [comment]
Wow, that diff really is bad! Thanks for the screenshot @psobolewskiPhD, that helps a lot :D

@aganders3 nice catch, thanks for the PR! Looks good to me.

**Review by brisvag — APPROVED** [review]

**Comment by aganders3** [comment]
Thanks for the reviews, and thanks @psobolewskiPhD for the clarification on the diff!

**Review by psobolewskiPhD — APPROVED** [review]
Everything looks good to me.
Edit: sorry didn't know how to do this properly—still learning!

**Comment by brisvag** [comment]
@aganders3 we have some merge conflicts, do you want to have a go at them?

**Comment by aganders3** [comment]
Thanks for the heads-up. The only difference I found was that `self._dims_displayed` became `self._slice_input.displayed`. I will request re-review if/when CI passes.

**Comment by andy-sweet** [comment]
As @psobolewskiPhD [mentioned in Zulip](https://napari.zulipchat.com/#narrow/stream/215289-release/topic/0.2E4.2E17/near/303664121), this is fairly niche, so I'm going to remove the v0.4.17 milestone from this to keep the RC changes small.

Thanks again for the fix @aganders3!
